octopoes/octopoes/api/models.py

```diff
@@ -59,18 +59,21 @@ class ScanProfileDeclaration(BaseModel):
     valid_time: datetime
 
 
+ValidatedOOIType = Annotated[ConcreteOOIType, Field(discriminator="object_type")]
```

Comment:
It took me quite a long time to figure out what was going wrong when I changed the `object_type` from the base OOI form `Literal["OOI"]` to `str` to fix typing conflicts with all the subclasses. What happened was that it created an IPAddress object instead of an IPAddressv4 object because the IPAddress suddenly also matched because `object_type` wasn't `Literal["OOI"]` anymore.

Using [discriminated unions](https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions) is actually a better way to do this matching. Pydantic will only try to validate against the type that matches the object_type instead of all the types. This is more efficient and any validation error will only about that type and not all every other type it would otherwise try. This doesn't work with abstract types that don't define `object_type` so I had to create the `ConcreteOOIType` type that remove those.

Reply:
This is indeed a huge change since v2. Discriminated unions weren't available before, hence the big union of concrete types. In v2 unions are also handled differently now.

It's something I've always wanted to do as soon as possible after the migration, but found it a bit risky. I'm not sure about the name yet, but I'm happy you picked this up. (Perhaps we could rename the old *OOIType* to something like *OOIBaseType* or *OOIGeneralType*, and make your suggested discriminator type *OOIType*)

Reply:
Discriminated unions were already available in 1.9 that was released in december 2021: https://docs.pydantic.dev/latest/changelog/#v190-2021-12-31

Reply:
Yep you're right. I meant to say that the way we handled these types was implemented even before 1.9, and after that we didn't implement discriminated unions yet.